### PR TITLE
Fixing NullReferenceException inside GenerateTypeCodeFixProvider when System.Action is missing

### LINKED ISSUE #22293: 'GenerateTypeCodeFixProvider' encountered an error and has been disabled
**Version Used**: 
Visual Studio Community 2017 v15.3.5

**Steps to Reproduce**:

1. File -> New -> Windows Classic Desktop Console App (.NET Framework) targeting .NET Framework 3.0
2. In Program.cs paste:
```csharp
using System;
using System.Threading;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            Task task = new Task(Display);
            task.Start();
        }

        static void Display()
        {
        }
    }
}
```
3. Over `new Task(Display)` choose 'Show potential fixes'

**Expected Behavior**:
'Show potential fixes' should suggest me to update to .NET Framework 4+ to use Task or didn't appears at all.

**Actual Behavior**:
Message "'GenerateTypeCodeFixProvider' encountered an error and has been disabled" apeears at the top of code editor window.

Pressing show stack trace gives:
```
System.NullReferenceException : Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.CSharp.Extensions.ITypeSymbolExtensions.GenerateTypeSyntax(INamespaceOrTypeSymbol symbol,Boolean nameSyntax)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.FieldGenerator.GenerateFieldDeclaration(IFieldSymbol field,CodeGenerationDestination destination,CodeGenerationOptions options)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.FieldGenerator.AddFieldTo(TypeDeclarationSyntax destination,IFieldSymbol field,CodeGenerationOptions options,IList`1 availableIndices)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.CSharpCodeGenerationService.AddField[TDeclarationNode](TDeclarationNode destination,IFieldSymbol field,CodeGenerationOptions options,IList`1 availableIndices)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.UpdateDestination[TDeclarationNode](IList`1 availableIndices,CodeGenerationOptions options,TDeclarationNode currentDestination,ISymbol member,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.AddMembersToAppropiateLocationInDestination[TDeclarationSyntax](TDeclarationSyntax destination,IEnumerable`1 members,IList`1 availableIndices,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.AddMembers[TDeclarationNode](TDeclarationNode destination,IEnumerable`1 members,IList`1 availableIndices,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.AddMembers[TDeclarationNode](TDeclarationNode destination,IEnumerable`1 members,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.NamedTypeGenerator.GenerateNamedTypeDeclaration(ICodeGenerationService service,INamedTypeSymbol namedType,CodeGenerationDestination destination,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.NamedTypeGenerator.AddNamedTypeTo(ICodeGenerationService service,NamespaceDeclarationSyntax destination,INamedTypeSymbol namedType,CodeGenerationOptions options,IList`1 availableIndices,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.CSharpCodeGenerationService.AddNamedType[TDeclarationNode](TDeclarationNode destination,INamedTypeSymbol namedType,CodeGenerationOptions options,IList`1 availableIndices,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.UpdateDestination[TDeclarationNode](IList`1 availableIndices,CodeGenerationOptions options,TDeclarationNode currentDestination,ISymbol member,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.AddMembersToAppropiateLocationInDestination[TDeclarationSyntax](TDeclarationSyntax destination,IEnumerable`1 members,IList`1 availableIndices,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.AddMembers[TDeclarationNode](TDeclarationNode destination,IEnumerable`1 members,IList`1 availableIndices,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.AddMembers[TDeclarationNode](TDeclarationNode destination,IEnumerable`1 members,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.NamespaceGenerator.GenerateNamespaceDeclaration(ICodeGenerationService service,INamespaceSymbol namespace,CodeGenerationOptions options,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.NamespaceGenerator.AddNamespaceTo(ICodeGenerationService service,CompilationUnitSyntax destination,INamespaceSymbol namespace,CodeGenerationOptions options,IList`1 availableIndices,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CodeGeneration.CSharpCodeGenerationService.AddNamespace[TDeclarationNode](TDeclarationNode destination,INamespaceSymbol namespace,CodeGenerationOptions options,IList`1 availableIndices,CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.<>c__DisplayClass54_0.<AddNamespaceAsync>b__0(SyntaxNode t,CodeGenerationOptions opts,IList`1 ai,CancellationToken ct)
   at async Microsoft.CodeAnalysis.CodeGeneration.AbstractCodeGenerationService.GetEditAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.GenerateType.AbstractGenerateTypeService`6.Editor.GetGenerateInNewFileOperationsAsync[TService,TSimpleNameSyntax,TObjectCreationExpressionSyntax,TExpressionSyntax,TTypeDeclarationSyntax,TArgumentSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.GenerateType.AbstractGenerateTypeService`6.Editor.GetOperationsAsync[TService,TSimpleNameSyntax,TObjectCreationExpressionSyntax,TExpressionSyntax,TTypeDeclarationSyntax,TArgumentSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.GenerateType.AbstractGenerateTypeService`6.GenerateTypeCodeAction.ComputeOperationsAsync[TService,TSimpleNameSyntax,TObjectCreationExpressionSyntax,TExpressionSyntax,TTypeDeclarationSyntax,TArgumentSyntax](<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.ComputePreviewOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.CodeActions.CodeAction.GetPreviewOperationsAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedAction.GetPreviewResultAsync(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Editor.Implementation.Suggestions.SuggestedActionWithNestedFlavors.<>c__DisplayClass11_0.<GetPreviewAsync>b__0(<Unknown Parameters>)
   at async Microsoft.CodeAnalysis.Extensions.IExtensionManagerExtensions.PerformFunctionAsync[T](<Unknown Parameters>)
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

### PR DESCRIPTION
**Pull request by Neme12**
fixes #22293 

the first commit fixes the bug, the rest should just be cleanup

### CODE REVIEW

File: src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs [hunk 1]
```diff
@@ -488,58 +488,31 @@ public static bool IsUnsafe(this ISymbol member)
...
-                    // +1 for the return type.
-                    var count = extensionUsedAsInstance ? method.Parameters.Length - 1 : method.Parameters.Length;
-                    var skip = extensionUsedAsInstance ? 1 : 0;
-                    var functionName = "System.Func`" + (count + 1);
-                    var functionType = compilation.GetTypeByMetadataName(functionName);
-
-                    if (functionType != null)
+                    var types = method.Parameters
+                        .Skip(skip)
+                        .Select(p => p.Type ?? compilation.GetSpecialType(SpecialType.System_Object));
```
Comment (Neme12) [inline src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs]: can this actually ever be null? is it possible this check was redundant?

the tests pass even if i remove this

Comment (Neme12) [reply src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs]: it's in the previous code too, but I'm not sure if it's actually useful...
on the other hand I don't want to fix one source of NullReferenceException and introduce another one :D

File: src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs [hunk 2]
```diff
@@ -4875,6 +4875,57 @@ public Goo(StackOverflowException e)
...
+    }}
+
+    {returnType} Method()
+    {{
+    }}
+}}
+
+internal class Class
+{{
+    private global::System.Object method;
```
Comment (Neme12) [inline src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs]: this must be caused by CommonReferences="false" for some reason... in VS with .NET 3.0 (as specified in the issue) it just uses `object`

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs]: Yup.

**Comment by Neme12** [comment]
retest windows_coreclr_debug_prtest  please

**Comment by DustinCampbell** [comment]
@sharwell : Pinging you directly here since the issue this fixes is assigned to you.

**Comment by sharwell** [comment]
Thanks @Neme12 😄

**Pull request merged**